packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,274 @@
+): UserAction[] => {
+  const mappedUserActions = rawActions
+    .map((rawAction) => {
+      const actionData = JSON.parse(rawAction.value);
+      return {
+        actionType: rawAction.name as UserActionType,
+        timestamp: rawAction.time,
+        eventData: {
+          type: actionData.event_type,
+          value: actionData.event_value,
+        },
+        cause: actionData.cause,
+        searchHub: actionData.origin_level_1,
+        document: {
+          title: actionData.title,
+          uriHash: actionData.uri_hash,
+          // eslint-disable-next-line @cspell/spellchecker
+          contentIdKey: actionData.c_contentidkey,
+          // eslint-disable-next-line @cspell/spellchecker
+          contentIdValue: actionData.c_contentidvalue,
+        },
+        query: actionData.query_expression,
+      };
+    })
+    .sort((a, b) => Number(b.timestamp) - Number(a.timestamp));
+
+  return mappedUserActions;
+};
+
+/**
+ * Splits the actions into sessions and build the timeline.
+ * @param actions {UserAction[]} - The actions to split
+ * @param ticketCreationDate {string} - The ticket creation date
+ * @returns {UserActionTimeline}
+ */
+export const splitActionsIntoTimelineSessions = (
+  actions: UserAction[],
+  ticketCreationDate: string
+): UserActionTimeline => {
+  const returnTimeline: UserActionTimeline = {
+    precedingSessions: [],
+    session: undefined,
+    followingSessions: [],
+    caseSessionFound: false,
+  };
+
+  let currentSession: UserSession = {
+    start: actions[0].timestamp,
+    end: actions[0].timestamp,
+    actions: [],
+  };
+
+  actions.forEach((action) => {
+    if (isActionWithin30Minutes(action, currentSession.end)) {
+      currentSession.actions.push(action);
+      currentSession.start = action.timestamp;
+      return;
+    }
+
+    if (sessionHasTicketCreation(currentSession, ticketCreationDate)) {
+      returnTimeline.session = currentSession;
+      returnTimeline.caseSessionFound = true;
+      return;
+    }
+
+    if (sessionIsBeforeCaseCreation(currentSession, ticketCreationDate)) {
+      returnTimeline.precedingSessions.push(currentSession);
+      return;
+    }
+
+    returnTimeline.followingSessions.push(currentSession);
+
+    currentSession = {
+      start: action.timestamp,
+      end: action.timestamp,
+      actions: [action],
+    };
+  });
+
+  // Inserting ticket creation action in current session
+  const caseCreationSessionActions: UserAction[] | [] =
+    insertTicketCreationActionInSession(
+      returnTimeline.session,
+      ticketCreationDate
+    );
+
+  const caseCreationSession = createCaseCreationSession(
+    caseCreationSessionActions
+  );
+  returnTimeline.session = caseCreationSession;
+
+  return returnTimeline;
+};
+
+/**
+ * Checks if an action is within 30mins of the previous action.
+ * @param action {UserAction} - The current action
+ * @param previousTimestamp {string} - The timestamp of the previous action
+ * @returns {boolean}
+ */
```

Comment:
I don't understand why this needs to be different than all other sessions?
If there is a session existing to include the case creation timestamp, you can just insert the caseCreation event in it.

If there's none, then you just create a new session with only the caseCreation event in it?